torch/distributed/distributed_c10d.py

```diff
@@ -1380,6 +1380,10 @@ def destroy_process_group(group: Optional[ProcessGroup] = None):
     if _world.pg_map.get(pg, None) is None:
         raise RuntimeError("Invalid process group specified")
 
+    # prevent Python Interpreter to exit before all hooks are fired
+    if pg.name().lower() == "nccl":
+        pg._wait_for_pending_works()
```

Comment:
It's likely that Python interpreter will exit before Cpp destructs PG instance and joins the watchdog thread. As a result, the Python hooks fired in watchdog thread will crash when attempting to grab GIL. So we basically have two options:

1. revive Python Interpreter if necessary before running hook
2. keep the interpreter alive until all hooks are fired. 

This PR attempts to follow options 2. Let me know if we prefer a different solution. 

Let me add more comments in code to explain this.

Reply:
A safer design would be to have all watchdog threads enqueue events (and not work objects) to a MPSC queue and then have a python thread dequeueing from them.

This would further improve isolation from callback issues.

Reply:
Hey @kumpera,  you mean start a thread in Python and call into C++ to release GIL? 

The current version spins up a dedicated thread from C++ to handle hooks (in parallel to NCCL watchDog). This PR is getting a bit too long. I'd prefer introduce more features in followup PRs if that's OK to you.